node/test/Infrastructure/Generator.hs

```diff
@@ -31,22 +67,213 @@ genChainUsingModel GeneratorModel{..} =
+    , initialSeed = Seed 0
+    , minFee = const 0 -- TODO: QUESTION: what properties should 'minFee' satisfy?
+    , initTransactions = abstractInitTs
+    -- TODO: how do we want to determine the bootstrap stakeholders?
+    , bootstrapStakeholders = Set.fromList $ toList addrs
+    }
+
+-- | Implementation of the slot leader function, called @sl@ in section 7.2.
+--
+-- This function relies on a /number generator/ function, which at the moment
+-- does not have /good randomness properties/. Furthermore, correctness of this
+-- function relies on the total stake (that is held by all the agents) being
+-- greater than zero (otherwise the number generator function is not defined).
+--
+mSlotLeader :: Seed -> StakeDistribution Addr -> SlotId -> Addr
+mSlotLeader s0 (StakeDistribution d) (SlotId n) = addrs !! idx
+  where
+    -- | List of addresses, which are taken from the support of 'd'.
+    addrs :: [Addr]
+    addrs = Set.toList (fSupport d)
+
+    -- | Index of the first cumulative stake bigger than 'un'.
+    --
+    -- NOTE: here is an example where it'd make sense not to have the @Stake@
+    -- represented as an 'Int'. We could have used the 'Data.List.find'
+    -- function, and still have everything type-check.
+    idx :: Int
+    idx = fromMaybe err $ findIndex (un <) cStake
+
+    -- | Cumulative sum of stakes of each address.
+    cStake :: [Int]
+    cStake = scanl' (+) 0 $ map (getSum . fApply d) addrs
+
+    -- | Stake held by all the agents.
+    totalStake :: Int
+    Sum totalStake = fSum d
+
+    -- | Value between 0 and the total stake that we use to choose a random
+    -- agent.
+    --
+    -- TODO: we need to assert that the '0 < totalStake'.
+    un :: Int
+    (un, _) = seeds s0 totalStake !! n
+
+    -- | Note that we should be able to find an index @i@ such that @un <
+    -- cStake !! i@, since @un < cStake !! (length cStake - 1) == totalStake@.
+    --
+    -- TODO: consider using liquid Haskell to actually verify this!
+    err =  error $  "The impossible has happened: "
+                 <> "could not find an index that accumulates enough stake."
+
+-- | Sequence of random number and seeds.
+seeds :: Seed -> Int -> [(Int, Seed)]
+seeds _ 0  = []
+seeds s0 n = (n, s0):[ numGen n sn | (_, sn) <- seeds s0 n ]
+
+-- | Implementation of the number generator function as described in Section
+-- 7.2 (called @f@).
+--
+-- Here the 'Int' values represent stake values.
+--
+-- Ideally `numGen` should be surjective so that each stakeholder could be
+-- selected by some random seed.
+--
+-- Note that this function is not defined when @i == 0@
+numGen :: Int  -> Seed -> (Int, Seed)
+-- Some dummy implementation for now...
+numGen i (Seed s) = (s `mod` i, Seed (s + 1))
+
+-- | For now we return the same seed. Ideally we'd need to perform an SCC
+-- computation, which uses the stake distribution and the VSS payloads (which
+-- should be tracked in the state).
+--
+mCurrentSeed :: TransState GivenHash -> Seed
```

Comment:
What is the `m` prefix in this and other functions here?

Reply:
Seeing all functions with a common prefix suggests there's a reason for it, yet this wasn't documented anywhere, hence my question.